### ISSUE TITLE
[Proposal] Update Illuminate\Support\Str::snake() to handle strings containing numbers

### ISSUE THREAD

**Issue by syntacticNaCl** [laravel/ideas#1514]
I am currently using the illuminate/support subtree split in one of my projects and I saw that repo was readonly so I am assuming this is where I would post an issue for it. If not please let me know.

I recently ran across an issue where I was needing to convert a string to snake case like `testProp1` to `test_prop_1`. I noticed that `Illuminate\Support\Str::snake()` does not do the replacement on numbers in the conversion. It would be great to have support for this without having to process the string again after snake casing it.

This would be a very simple fix on https://github.com/laravel/framework/blob/6b0268be913e33cef0f122049bcd9da2578db1e5/src/Illuminate/Support/Str.php#L457:
```php
$value = static::lower(preg_replace('/(.)(?=[A-Z])/u', '$1'.$delimiter, $value));
```
to 
```php
$value = static::lower(preg_replace('/(.)(?=[A-Z0-9])/u', '$1'.$delimiter, $value));
```

**Comment by mfn** [comment]
👎 because it changes the behaviour on a wildly used function.

Either create dedicated helper in your application or yo need to propose a new name, too.

**Comment by donnysim** [comment]
And also `/(.)(?=[A-Z0-9])/u` would translate `testProp12` into `test_prop_1_2`.

**Comment by syntacticNaCl** [comment]
Thanks @mfn. 
The reasoning for posting this issue was all based on my complete misunderstanding of the purpose of `snakeAttributes` in https://github.com/laravel/framework/blob/495ec792513d849da8926badac5535f18daa2c53/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L71. I thought that this allowed for you to use camelCase properties on the model but it looks like that is not the case. 

In my current implementation we are using `illuminate/database` and `illuminate/support` to use within our existing code base. In an attempt to stick with our current paradigms we are using getters and setters instead of accessing the properties directly on a model, which allows us to enforce strict return typing, among other things. I say this because my specific issue is that I had some column names in a table with numbers, not my favorite but in this case it is unavoidable. I encountered an issue when attempting to camel case the properties in a `getAttribute` method that we are using in a class that extends `Eloquent\Model`. The method looks something like this:
```php
public function getAttribute($key)
{
        $snakeKey = Str::snake($key);

        $value = parent::getAttribute($snakeKey);

        if (array_key_exists($key, $this->relations)) {
            $value = parent::getAttribute($key);
        }
}
```

When attempting to use `Str::snake()` on a property (which maps to a column in our DB), it fails to snake case the numbers which makes it inaccessible in the `attributes` array and returns a NULL. This is an abnormal use case and one that does not seem to be an issue in the Laravel framework, again due to `snakeAttributes = false` not doing what I thought 😄. The only references to this `snakeAttributes` static property seems to be when dealing with mutators (https://github.com/laravel/framework/blob/495ec792513d849da8926badac5535f18daa2c53/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L255 and https://github.com/laravel/framework/blob/495ec792513d849da8926badac5535f18daa2c53/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L1222) and from what I can tell, would never be an issue since the developer explicitly sets what property the mutator mutates.

Although it would be nice for the `Str::snake()` method to support this, I understand that it is a widely used method and one in which its use case would not cause a side effect like this. However, I do think creating a method `Str::snakeAlphanumeric()` or something to that effect would be useful. Any thoughts on that?

**Comment by syntacticNaCl** [comment]
thanks @donnysim. I actually came into work to see the same comment from a co-worker on my work PR :smile:. It would need to be more like this:
```php
$value = static::lower(preg_replace('/(.)(?=[A-Z])/u', '$1'.$delimiter, $value));
$value = preg_replace('/([a-z])(?=\d)/u', '$1'.$delimiter, $value);
```
I am still pretty noobish with regex so I am sure there is a better way to do it.

**Comment by mfn** [comment]
> In my current implementation we are using `illuminate/database` and `illuminate/support` to use within our existing code base. In an attempt to stick with our current paradigms we are using getters and setters instead of accessing the properties directly on a model, which allows us to enforce strict return typing, among other things.

I can't comment much on your specific case, but maybe it helps what my experience regarding this is.

I'm a user of laravel-framework and a non-laravel-framework project which uses those two packages.

In Laravel project itself I value/enforce types as "as good as it gets" via laravel-ide-helper and phpstan.

Initially I sorely missed this from the non-laravel-framework but breaking the attribute accessor convention was a non-starter for my team, as everyone expects it to behave/work the same in Laravel. Notwithstanding the additional boilerplate…

TL;DR: I did hack around laravel-ide-helper to give me the virtual `@properties` based on DB introspection even on the non-laravel-framework project and it works as well as within the actual Laravel framework.

I've detailed my experience how I did integrate it in https://github.com/barryvdh/laravel-ide-helper/issues/756

HTH!

**Comment by syntacticNaCl** [comment]
Thanks @mfn. These are very cool packages and although it might not work for a majority of our team, I will definitely pass this on for users of PHPStorm and Sublime.

If there doesn't seem to be any want for another method like `snakeAlphanumeric()` then I am fine with closing this issue. For now we will just stick with a custom method that handles this from the `getAttributes` and `setAttributes` that we have overwritten from the `HasAttribute` trait. Thanks again for your suggestions!
